docs/proposals-accepted/202301-distributed-query-execution.md

```diff
@@ -0,0 +1,155 @@
+---
+type: proposal
+title: Distributed Query Execution
+status: accepted
+owner: fpetkovski
+menu: proposals-accepted
+---
+
+## 1 Related links/tickets
+
+* https://github.com/thanos-io/thanos/pull/5250
+* https://github.com/thanos-io/thanos/pull/4917
+* https://github.com/thanos-io/thanos/pull/5350
+* https://github.com/thanos-community/promql-engine/issues/25
+
+## 2 Why
+
+Thanos Queriers currently need to pull in all data from Stores in memory before they can start evaluating a query. This has a large impact on the used memory inside a single querier, and drastically increases query execution latency.
+
+Even when a Querier is connected to other Queriers, it will still pull raw series instead of delegating parts of the execution to its downstreams. This document proposes a mode in the Thanos Querier where it will dispatch parts of the execution plan to different, independent Queriers.
```

Comment:
I am trying to understand how much this helps for query execution. Even if the central querier delegates part of execution to other queriers and we did some downsampling to fetch less data, it is only beneficial for central querier to use less memory as far as I can see.
For remote queriers and stores, we still need to pull all raw data into memory and evaluate locally. This sounds worse if we are just using one querier.

Reply:
Maybe I missed something. I can see it helps if we are using a layer architecure of queriers like you have a central querier and multiple sub queriers. If there is only one layer I feel nothing changed.

Reply:
Yes, layering is the proposed way of running the distributed engine. This is also how Mimir's sharidng works, it's just that there's an embedded engine in query frontend. So layering still exists, it's just not so explicit. A layered setup is easy to achieve, for example with running one querier for N store gateways, or even having one querier per store gateway.

I don't think having shared storage for all queriers and trying to do some kind of filtering is a scalable approach, I tried to explain this a bit in the first section. You might have a different experience, but we notice a plateau with vertical sharding after 4 or 5 shards. Query latency also becomes far less predictable and we see timeouts from stores. I suspect this is because there is a lot of contention and load in stores with both vertical and horizontal sharding since each subquery goes to the same stores. With this approach, we can scale horizontally indefinitely because load will always be spread.

> In this case, central queriers can fetch less data with the help of fetching downsampled and aggregated data.

The bigger advantage is that subqueriers will execute queries in parallel on a much smaller subset of the data.